File: email-aggregator-go/src/store/metadata.go
// Package store 存储层：元数据仓储接口与 PoC 内存实现（ADR-003）。
// 生产按 account_id 分片；真实 PG 实现见 integration.PgMetadataStore（//go:build integration）。
package store

import (
	"fmt"
	"sync"

	"email-aggregator-go/src/model"
	"email-aggregator-go/src/tenant"
)

// ns 租户命名空间前缀：把所有数据面键（邮件幂等键 / 账户 / 游标）都收敛到 tenant 之下，
// 实现「逻辑多租户」隔离边界（ADR-009）。单租户下 tenantID=default，等价于改造前的 account 隔离。
func ns(tenantID, key string) string { return tenantID + "\x00" + key }

// MetadataStore 元数据仓储接口（ADR-009：tenantID 全链路首参强制透传）
type MetadataStore interface {
	UpsertMail(tenantID string, m model.CanonicalMail) error
	GetMail(tenantID, accountID, idempotencyKey string) (*model.CanonicalMail, error)
	ListMails(tenantID, accountID, folder string, limit int) ([]model.CanonicalMail, error)
	ListAccounts(tenantID string) ([]string, error)
	Count(tenantID, accountID string) (int, error)
	SetRead(tenantID, accountID, id string, read bool) error
	DeleteMail(tenantID, accountID, id string) error
	UnreadCount(tenantID, accountID string) (int, error)
	CursorStore // 同步游标持久化：重启后增量续拉，避免重复全量
}

// CursorStore 同步游标持久化（按 tenant+account+folder）
type CursorStore interface {
	GetCursor(tenantID, accountID, folder string) (model.SyncCursor, error)
	PutCursor(tenantID, accountID, folder string, cursor model.SyncCursor) error
}

// InMemoryMetadataStore 演示/单测用内存实现（幂等去重）
type InMemoryMetadataStore struct {
	mu      sync.RWMutex
	byKey   map[string]model.CanonicalMail
	byAcct  map[string][]model.CanonicalMail
	cursors map[string]model.SyncCursor
}

// NewInMemoryMetadataStore 构造
func NewInMemoryMetadataStore() *InMemoryMetadataStore {
	return &InMemoryMetadataStore{
		byKey:   map[string]model.CanonicalMail{},
		byAcct:  map[string][]model.CanonicalMail{},
		cursors: map[string]model.SyncCursor{},
	}
}

// UpsertMail 幂等写入（按 tenant+id 去重；强制盖上租户归属，保证隔离边界）
func (s *InMemoryMetadataStore) UpsertMail(tenantID string, m model.CanonicalMail) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	tenantID = tenant.Resolve(tenantID)
	m.TenantID = tenantID
	key := ns(tenantID, m.ID) // 简化：以邮件 ID 作幂等键（真实用 tenant:accountId:mailId 或 provider 稳定 ID）
	if _, ok := s.byKey[key]; ok {
		return nil // 已存在，幂等跳过
	}
	s.byKey[key] = m
	s.byAcct[ns(tenantID, m.AccountID)] = append(s.byAcct[ns(tenantID, m.AccountID)], m)
	return nil
}

// GetMail 按租户+账户+幂等键取邮件
func (s *InMemoryMetadataStore) GetMail(tenantID, accountID, key string) (*model.CanonicalMail, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	tenantID = tenant.Resolve(tenantID)
	m, ok := s.byKey[ns(tenantID, key)]
	if !ok || m.AccountID != accountID {
		return nil, nil
	}
	return &m, nil
}

// ListMails 列出租户内某账户某文件夹邮件（按时间倒序，截取 limit）
func (s *InMemoryMetadataStore) ListMails(tenantID, accountID, folder string, limit int) ([]model.CanonicalMail, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	tenantID = tenant.Resolve(tenantID)
	all := s.byAcct[ns(tenantID, accountID)]
	out := make([]model.CanonicalMail, 0, len(all))
	for i := len(all) - 1; i >= 0; i-- {
		if all[i].Folder == folder {
			out = append(out, all[i])
		}
	}
	if limit > 0 && len(out) > limit {
		out = out[:limit]
	}
	return out, nil
}

// Count 租户内账户邮件数
func (s *InMemoryMetadataStore) Count(tenantID, accountID string) (int, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	tenantID = tenant.Resolve(tenantID)
	return len(s.byAcct[ns(tenantID, accountID)]), nil
}

// ListAccounts 列出租户内当前已写入邮件的所有账户 ID（按首次出现顺序，稳定可演示）。
func (s *InMemoryMetadataStore) ListAccounts(tenantID string) ([]string, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	tenantID = tenant.Resolve(tenantID)
	seen := map[string]struct{}{}
	out := make([]string, 0)
	for _, m := range s.byKey {
		if m.TenantID != tenantID {
			continue
		}
		if _, ok := seen[m.AccountID]; ok {
			continue
		}
		seen[m.AccountID] = struct{}{}
		out = append(out, m.AccountID)
	}
	return out, nil
}

// SetRead 设置邮件已读/未读状态（CRUD 生命周期）。
func (s *InMemoryMetadataStore) SetRead(tenantID, accountID, id string, read bool) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	tenantID = tenant.Resolve(tenantID)
	m, ok := s.byKey[ns(tenantID, id)]
	if !ok || m.AccountID != accountID {
		return fmt.Errorf("mail not found")
	}
	m.Read = read
	s.byKey[ns(tenantID, id)] = m
	acct := s.byAcct[ns(tenantID, accountID)]
	for i := range acct {
		if acct[i].ID == id {
			acct[i].Read = read
			break
		}
	}
	return nil
}

// DeleteMail 删除邮件（软删除演示：从内存索引移除）。
func (s *InMemoryMetadataStore) DeleteMail(tenantID, accountID, id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	tenantID = tenant.Resolve(tenantID)
	m, ok := s.byKey[ns(tenantID, id)]
	if !ok || m.AccountID != accountID {
		return fmt.Errorf("mail not found")
	}
	delete(s.byKey, ns(tenantID, id))
	acct := s.byAcct[ns(tenantID, accountID)]
	out := make([]model.CanonicalMail, 0, len(acct))
	for _, x := range acct {
		if x.ID != id {
			out = append(out, x)
		}
	}
	s.byAcct[ns(tenantID, accountID)] = out
	return nil
}

// UnreadCount 租户内账户未读邮件数。
func (s *InMemoryMetadataStore) UnreadCount(tenantID, accountID string) (int, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	tenantID = tenant.Resolve(tenantID)
	n := 0
	for _, m := range s.byAcct[ns(tenantID, accountID)] {
		if !m.Read {
			n++
		}
	}
	return n, nil
}

// GetCursor 取游标
func (s *InMemoryMetadataStore) GetCursor(tenantID, accountID, folder string) (model.SyncCursor, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	tenantID = tenant.Resolve(tenantID)
	c, ok := s.cursors[ns(tenantID, accountID)+"|"+folder]
	if !ok {
		return model.SyncCursor{}, nil
	}
	return c, nil
}

// PutCursor 存游标
func (s *InMemoryMetadataStore) PutCursor(tenantID, accountID, folder string, cursor model.SyncCursor) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	tenantID = tenant.Resolve(tenantID)
	s.cursors[ns(tenantID, accountID)+"|"+folder] = cursor
	return nil
}

// 真实 PG 实现见 integration/integration.go 中的 PgMetadataStore（//go:build integration）。
// 替换要点（与 TS 版一致）：
//
//	INSERT ... ON CONFLICT (idempotency_key) DO NOTHING;
//	游标表 account_sync_cursor(account_id, folder, cursor_json)；按 account_id 分片（Citus）。

File: email-aggregator/src/store/metadata.ts
/**
 * 元数据仓储。元数据（发件人/主题/时间/文件夹/游标）存 PostgreSQL，
 * 后续按 tenant_id + account_id 分片（ADR-009）。PoC 以 InMemoryMetadataStore 实现，
 * PgMetadataStore 为真实集成 stub（接口一致）。
 *
 * 与 Go 实现对齐：email-aggregator-go/src/store/metadata.go（tenantId 首参 + Resolve 收敛）。
 */

import type { CanonicalMail, FolderName, SyncCursor } from "../model/canonical.ts";
import type { CursorStore } from "../sync/orchestrator.ts";
import { Resolve } from "../tenant/tenant.ts";

export interface MetadataStore {
  upsertMail(m: CanonicalMail): Promise<void>;
  getMail(tenantId: string, accountId: string, idempotencyKey: string): Promise<CanonicalMail | null>;
  listMails(tenantId: string, accountId: string, folder: FolderName, limit: number): Promise<CanonicalMail[]>;
  count(tenantId: string, accountId: string): Promise<number>;
}

export class InMemoryMetadataStore implements MetadataStore, CursorStore {
  // `${tenantId}|${idempotencyKey}` -> mail（防跨租户同键碰撞）
  private byKey = new Map<string, CanonicalMail>();
  // `${tenantId}|${accountId}` -> mails（租户隔离的主索引）
  private byAccount = new Map<string, CanonicalMail[]>();
  // `${tenantId}|${accountId}|${folder}` -> cursor
  private cursors = new Map<string, SyncCursor>();

  async upsertMail(m: CanonicalMail): Promise<void> {
    const tid = Resolve(m.tenantId);
    const mWithTid: CanonicalMail = { ...m, tenantId: tid };
    const k = `${tid}|${m.idempotencyKey}`;
    this.byKey.set(k, mWithTid);
    const acc = `${tid}|${m.accountId}`;
    const list = this.byAccount.get(acc) ?? [];
    if (!list.find((x) => x.idempotencyKey === m.idempotencyKey)) list.push(mWithTid);
    this.byAccount.set(acc, list);
  }

  async getMail(tenantId: string, accountId: string, idempotencyKey: string): Promise<CanonicalMail | null> {
    const tid = Resolve(tenantId);
    const m = this.byKey.get(`${tid}|${idempotencyKey}`);
    return m && m.accountId === accountId ? m : null;
  }

  async listMails(tenantId: string, accountId: string, folder: FolderName, limit: number): Promise<CanonicalMail[]> {
    const tid = Resolve(tenantId);
    const list = this.byAccount.get(`${tid}|${accountId}`) ?? [];
    return list.filter((m) => m.folder === folder).slice(-limit).reverse();
  }

  async count(tenantId: string, accountId: string): Promise<number> {
    const tid = Resolve(tenantId);
    return (this.byAccount.get(`${tid}|${accountId}`) ?? []).length;
  }

  // ---- CursorStore 实现（落在元数据层，按租户+账户+文件夹持久化） ----
  async get(tenantId: string, accountId: string, folder: string): Promise<SyncCursor | null> {
    return this.cursors.get(`${Resolve(tenantId)}|${accountId}|${folder}`) ?? null;
  }
  async put(tenantId: string, accountId: string, folder: string, cursor: SyncCursor): Promise<void> {
    this.cursors.set(`${Resolve(tenantId)}|${accountId}|${folder}`, cursor);
  }
}

/**
 * @deprecated STUB — 未实现。仅作接口文档保留。
 * 真实 PG 适配器落地时替换为（按 tenant_id + account_id 分片）：
 *   INSERT ... ON CONFLICT (tenant_id, idempotency_key) DO NOTHING;
 *   -- 游标表 account_sync_cursor(tenant_id, account_id, folder, cursor_json)
 * Go 版真实实现见 integration/integration.go 中的 PgMetadataStore。
 */
export class PgMetadataStore implements MetadataStore {
  private dsn: string;
  constructor(dsn: string) {
    this.dsn = dsn;
    console.warn(`[PgMetadataStore] stub: 未连接 ${dsn}`);
  }
  async upsertMail(_m: CanonicalMail): Promise<void> {
    throw new Error("PgMetadataStore.upsertMail not implemented");
  }
  async getMail(_t: string, _a: string, _k: string): Promise<CanonicalMail | null> {
    return null;
  }
  async listMails(_t: string, _a: string, _f: FolderName, _l: number): Promise<CanonicalMail[]> {
    return [];
  }
  async count(_t: string, _a: string): Promise<number> {
    return 0;
  }
}
